### DISCUSSION TITLE
[Feature Request]: Use css variables instead of hardcoding colors

### DISCUSSION THREAD

**Discussion by thecashewtrader** [tailwindlabs/tailwindcss#5537] [Ideas]
# How does it work now?
Right now, if you edit the `tailwind.config.cjs` file and add, say, `theme-green: #a3be8c` to colors and use it as, say, `class="text-theme-green"`, the css which is generated will have this value hardcoded as rgb(a).
# How do you want it?
It'd be nice to have it so that it generates a
```
:root {
--theme-green: #a3be8c
}
```
and then uses `var(--theme-green)` instead of the hardcoded value
# Why?
This would make client side customisations with, say, Stylus much easier. Sorry if this is not in the scope of tailwindcss, I figured it'd be good to atleast have a discussion about it.
# EDIT: Fixed formatting

**Comment by cairin** [comment]
I'm fairly certain you can do this already.

In your main css file make sure your css variables come before tailwind's:

```css
:root {
  --theme-green: #a3be8c
}

@tailwind base;
@tailwind components;
@tailwind utilities;

```

Then in your tailwind config:

```javascript
module.exports = {
  theme: {
    extend: {
      colors: {
        'theme-green': 'var(--theme-green)'
      },
    },
  },
}
```
Remember to extend if it's not directly overriding an existing colour.
Now tailwind will use the variable `--theme-green' instead of a hard-coded value.

**Reply by thecashewtrader** [reply]
Wouldn't it be better if tailwind did this by default?
Because this would lead to code duplication, though minor.

**Reply by cairin** [reply]
> Wouldn't it be better if tailwind did this by default?

I'm not sure that it would.

From my understanding, Tailwind configures everything in javascript, and then compiles and injects the compiled css into the tailwind directives (`@tailwind base;` etc.) - replacing them with css based on your config.

So in order to achieve what you want, you would need to ensure that all those css variables are present as well. This potentially opens the tailwind config up to conflicts with other utilities that could lead to unexpected side affects (As apposed to just hard coding it).

For your example, Tailwind would by default produce the following class:

```css
.text-theme-green {
  color: #a3be8c;
}
```

However, using variables you get the following:
```css
:root {
  --theme-green: #a3be8c;
}

.text-theme-green {
  color: var(--theme-green);
}
```

I would imagine this would have negative affects on file size too.

> Because this would lead to code duplication, though minor.

I don't think this is the case. You're defining a variable in one place (your css file), and consuming it in another (your tailwind config). You're not redefining the css variable in the tailwind config.

**Reply by thecashewtrader** [reply]
Aight, understandable 😄

**Reply by thecashewtrader** [reply]
Also, I just noticed this is a "discussion" and not an issue, is there something I should do to achieve the equivalent of closing an issue?

**Reply by cairin** [reply]
> Also, I just noticed this is a "discussion" and not an issue, is there something I should do to achieve the equivalent of closing an issue?

I don't actually know. Sometimes you can mark a correct answer? But I think the repo owner has to set that up per topic.

**Reply by simonswiss** [reply]
Discussions live _forever_ – and only the "help 🙏" category has a "mark as correct answer" feature.

It's good to leave these open anyway, some conversations suddenly resurface or help someone else months down the road 👍

By the way - this video I made a while back is related to this topic. Some of the APIs in there may have changed slightly, but the content is still relevant:

https://www.youtube.com/watch?v=MAtaT8BZEAo
